karrot/activities/tests/test_activities_api.py

```diff
@@ -497,6 +500,95 @@ def test_export_ics_activities_as_group_member(self):
         self.assertEqual(response.status_code, status.HTTP_200_OK, response.data)
 
 
+class TestActivitiesWithRequiredRolesAPI(APITestCase):
+    def setUp(self):
+        self.member = UserFactory()
+        self.other_member = UserFactory()
+        self.group = GroupFactory(members=[self.member, self.other_member])
+        self.place = PlaceFactory(group=self.group, status='active')
+        self.approved_member = UserFactory()
+        self.group.groupmembership_set.create(
+            user=self.approved_member,
+            roles=[APPROVED],
+        )
+        self.activity = ActivityFactory(
+            place=self.place,
+            require_role=APPROVED,
+            max_open_participants=1,
+        )
+
+    def test_cannot_join_if_requires_role_and_none_without_role(self):
+        activity = ActivityFactory(
+            place=self.place,
+            require_role=APPROVED,
+            max_open_participants=0,
+        )
+        self.client.force_login(user=self.member)
+        response = self.client.post('/api/activities/{}/add/'.format(activity.id))
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+
+    def test_can_join_as_participant_without_role(self):
+        self.client.force_login(user=self.member)
+        response = self.client.post('/api/activities/{}/add/'.format(self.activity.id))
+        self.assertEqual(response.status_code, status.HTTP_200_OK, response.data)
+        participant = ActivityParticipant.objects.get(activity=self.activity, user=self.member)
+        self.assertEqual(participant.role, None)
+
+    def test_cannot_join_as_participant_without_role_if_full(self):
+        self.activity.add_participant(self.member, role=None)
+        self.client.force_login(user=self.other_member)
+        response = self.client.post('/api/activities/{}/add/'.format(self.activity.id))
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+
+    def test_can_join_as_normal_participant_if_has_role(self):
+        self.client.force_login(user=self.approved_member)
+        response = self.client.post('/api/activities/{}/add/'.format(self.activity.id), {'role': APPROVED})
+        self.assertEqual(response.status_code, status.HTTP_200_OK, response.data)
+        participant = ActivityParticipant.objects.get(activity=self.activity, user=self.approved_member)
+        self.assertEqual(participant.role, APPROVED)
+
+    def test_backwards_compatible_participants_api(self):
+        self.activity.add_participant(self.member, role=None)
+        self.activity.add_participant(self.approved_member, role=APPROVED)
+        self.client.force_login(user=self.member)
+        response = self.client.get('/api/activities/{}/'.format(self.activity.id))
+
+        # participants field only shows participants without roles user ids
+        self.assertEqual(response.data['participants'], [self.approved_member.id])
```

Comment:
This test no longer does what it says it does, as the existing frontend won't send the `role` parameter.